plugin/util/passwords.go

```diff
@@ -12,27 +13,65 @@ var (
 	minimumLengthOfComplexString = 8
 
 	PasswordComplexityPrefix = "?@09AZ"
-	MinimumPasswordLength    = len(PasswordComplexityPrefix) + minimumLengthOfComplexString
+	PwdFieldTmpl             = "{{PASSWORD}}"
 )
 
-func GeneratePassword(desiredLength int) (string, error) {
-	if desiredLength < MinimumPasswordLength {
-		return "", fmt.Errorf("it's not possible to generate a _secure_ password of length %d, please boost length to %d, though Vault recommends higher", desiredLength, MinimumPasswordLength)
+func GeneratePassword(formatter string, totalLength int) (string, error) {
+	if err := ValidatePwdSettings(formatter, totalLength); err != nil {
+		return "", err
+	}
+	pwd, err := generatePassword(totalLength)
+	if err != nil {
+		return "", err
+	}
+	if formatter == "" {
+		pwd = PasswordComplexityPrefix + pwd
+		return pwd[:totalLength], nil
+	}
+	return strings.Replace(formatter, PwdFieldTmpl, pwd[:lengthOfPassword(formatter, totalLength)], 1), nil
+}
+
+func ValidatePwdSettings(formatter string, totalLength int) error {
+	// Check for if there's no formatter.
+	if formatter == "" {
+		if totalLength < len(PasswordComplexityPrefix)+minimumLengthOfComplexString {
+			return fmt.Errorf("it's not possible to generate a _secure_ password of length %d, please boost length to %d, though Vault recommends higher", totalLength, minimumLengthOfComplexString)
+		}
+		return nil
 	}
 
+	// Check for if there is a formatter.
+	if lengthOfPassword(formatter, totalLength) < minimumLengthOfComplexString {
```

Comment:
For the case that a formatter is provided, I think there needs to be a check that `PwdFieldTmpl` is also provided at least once, otherwise passwords generated would simply be the formatter provided with no entropy.

Reply:
Whoops, typed too fast.. The check the down below :facepalm: